### PR TITLE
encoding/json: Add custom JSON package with build tag support for Sonic

### PR DESCRIPTION
**Pull request by shazbert**
# PR Description

This pull request introduces a custom JSON package located in encoding/json/ to facilitate swapping between the default Go encoding/json package and the sonic library based on build tags. 

Adds:
* json.go (implementation using encoding/json).
* sonic.go (default implementation using sonic library).

Import Updates:

All imports to the standard encoding/json package have been updated to the new custom package github.com/thrasher-corp/gocryptotrader/encoding/json.

Added basic benchmarks for both the default encoding/json and the sonic implementation to evaluate *hot-swap* performance.

## Type of change

Please delete options that are not relevant and add an `x` in `[]` as item is complete.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How has this been tested

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration and
also consider improving test coverage whilst working on a certain feature or package.

- [ ] go test ./... -race
- [ ] golangci-lint run
- [ ] Test X

## Checklist

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation and regenerated documentation via the documentation tool
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally and on Github Actions with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Comment by shazbert** [comment]
Flagged this as medium because this is beast. Sorry for those that cant use it. 🙏

File: encoding/json/sonic.go [hunk 1]
```diff
@@ -0,0 +1,20 @@
+//go:build sonic && !(darwin || arm64)
```
Comment (gloriousCode) [inline encoding/json/sonic.go:1]: I'm not seeing anything that suggests mac or arm64 is incompatible with this. They may have added more support but I was questioning this at the time you raised the PR too.

File: encoding/json/sonic_benchmark_test.go [hunk 2]
```diff
@@ -0,0 +1,13 @@
+//go:build sonic && !(darwin || arm64)
+
+package json
+
+import "testing"
+
+// BenchmarkUnmarshal-16  1900803  631.8 ns/op
+// Usage: go test --tags=sonic -bench=BenchmarkUnmarshal -v
+func BenchmarkUnmarshal(b *testing.B) {
```
Comment (gloriousCode) [inline encoding/json/sonic_benchmark_test.go:9]: Only one of these is needed since they're the same test and doesn't matter what the build tag is
You can keep both to be explicit if you want, but if the build tag changes the json processor changes regardless

File: .github/workflows/tests.yml [hunk 3]
```diff
@@ -13,22 +13,52 @@ jobs:
             goarch: amd64
```
Comment (gloriousCode) [inline .github/workflows/tests.yml:13]: Unrelated to this line, would you consider a log on GCT start to say which one is enabled? Yes, you have build tags, but then it puts it in the logs of the application

Could add to the build tag files to call like `json.Version()` where its `fmt.Println("hello my name is Sonic and I gotta go fast, okay?")` as an example

**Review by gloriousCode — CHANGES_REQUESTED** [review]
image
Nice work on finding this. I like the way you've implemented it and makes it easy to experiment with out parsers in future with your groundwork

orig:
```
// BenchmarkUnmarshal-10    	  949106	      1065 ns/op	     816 B/op	      24 allocs/op
```
sonic:
```
// BenchmarkUnmarshal-10    	 1818868	       643.8 ns/op	     951 B/op	      19 allocs/op
```

File: .github/workflows/tests.yml [hunk 4]
```diff
@@ -13,22 +13,52 @@ jobs:
...
             psql: true
             skip_wrapper_tests: true
+            sonic: false
           - os: windows-latest
             goarch: amd64
             psql: true
             skip_wrapper_tests: true
+            sonic: false
+          - os: ubuntu-latest-sonic
```
[the remark this answers is not in the document]
Comment (shazbert) [reply .github/workflows/tests.yml]: > I think you should stick to the same style as the existing builds names with the false, false,false,true, would also just be fine with including sonic somewhere in the build name. Maybe its just not showing up below because its not merged?

That should have been reverted. I think. 

> Also, do we need two of each build? Would one extra one with sonic enabled suffice?

@thrasher- Confirming entire suite test or can just have one?

File: .github/workflows/tests.yml (hunk 3 above)
Comment (shazbert) [reply .github/workflows/tests.yml:13]: [needs checking](https://github.com/thrasher-corp/gocryptotrader/pull/1623/commits/1c9f9c2ee6adcd8c0c79932526602fe189b4979e)

File: encoding/json/sonic_benchmark_test.go (hunk 2 above)
Comment (shazbert) [reply encoding/json/sonic_benchmark_test.go]: [addressed](https://github.com/thrasher-corp/gocryptotrader/pull/1623/commits/3905b625d959af4a2784355f90f72630a4d041ed)

File: encoding/json/sonic.go (hunk 1 above)
Comment (shazbert) [reply encoding/json/sonic.go]: [fixed](https://github.com/thrasher-corp/gocryptotrader/pull/1623/commits/1697977b1ddaba0ea9a78cee01c156fcf607e7ec)

File: .github/workflows/tests.yml (hunk 4 above)
[the remark this answers is not in the document]
Comment (thrasher-) [reply .github/workflows/tests.yml]: Can target amd64 linux for now since that's our main deployment OS in use ATM

[the remark this answers is not in the document]
Comment (shazbert) [reply .github/workflows/tests.yml]: [done](https://github.com/thrasher-corp/gocryptotrader/pull/1623/commits/9cffc8f7a900b6186d2041412e1d4eef0099947e)

File: .golangci.yml [hunk 5]
```diff
@@ -117,6 +117,17 @@ linters:
 #   - zerologlint
 
 linters-settings:
+  depguard:
```
Comment (gloriousCode) [inline .golangci.yml:120]: This is a bit fun, nice work

File: README.md [hunk 6]
```diff
@@ -77,6 +77,7 @@ However, we welcome pull requests for any exchange which does not match this cri
 + WebGUI (discontinued).
 + Exchange HTTP mock testing. See [mock](/exchanges/mock/README.md).
 + Exchange multichain deposits and withdrawals for specific exchanges. See [multichain transfer support](/docs/MULTICHAIN_TRANSFER_SUPPORT.md).
++ **Sonic JSON Integration**: Swap between the default Go 'encoding/json' package and the sonic library using go build tags for optimized JSON handling `-tags=sonic` or by using the make file `make sonic`
```
Comment (gloriousCode) [inline README.md:80]: Please keep styling in line with the others (or update the others) and update the template as well.

File: encoding/json/json.go [hunk 7]
```diff
@@ -0,0 +1,23 @@
+//go:build !sonic
+
+package json
+
+import "encoding/json" //nolint:depguard // This is a wrapper package
```
Comment (gloriousCode) [inline encoding/json/json.go:5]: Please update comment. This PR has no readme of new functionality being introduced. Explaining that its here for the rest of the application to use the default go json package would be nice

File: encoding/json/common.go [hunk 8]
```diff
@@ -0,0 +1,13 @@
+package json
+
+import "encoding/json" //nolint:depguard // This is a wrapper package
```
Comment (gloriousCode) [inline encoding/json/common.go:3]: Update comment, but less detail needed than the other one

File: encoding/json/sonic.go [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+	NewDecoder = sonic.ConfigStd.NewDecoder
+	// MarshalIndent is like Marshal but applies Indent to format the output. See the "github.com/bytedance/sonic" documentation for MarshalIndent
+	MarshalIndent = sonic.ConfigStd.MarshalIndent
+	// Valid reports whether data is a valid JSON encoding. See the "github.com/bytedance/sonic" documentation for Valid
+	Valid = sonic.ConfigStd.Valid
+)
+
+func init() {
+	log.Printf("Using %s for JSON encoding\n", Version)
+}
```
Comment (gloriousCode) [inline encoding/json/sonic.go:31]: Makes more sense to have this in `common.go` because it doesn't output the default package
Also would recommend it being a function to call at a later stage. Only because its ugly! 😆 I don't mind if its not done, but it gets in front of our sweet banner
```
2025/01/28 10:40:52 Using encoding/json for JSON encoding

   ______        ______                     __        ______                  __
  / ____/____   / ____/_____ __  __ ____   / /_ ____ /_  __/_____ ______ ____/ /___   _____
 / / __ / __ \ / /    / ___// / / // __ \ / __// __ \ / /  / ___// __  // __  // _ \ / ___/
/ /_/ // /_/ // /___ / /   / /_/ // /_/ // /_ / /_/ // /  / /   / /_/ // /_/ //  __// /
\____/ \____/ \____//_/    \__, // .___/ \__/ \____//_/  /_/    \__,_/ \__,_/ \___//_/
                          /____//_/
GoCryptoTrader v0.1 arm64 go1.23.4 pre-release.
This version is pre-release and is not intended to be used as a production ready trading framework or bot - use at your own risk.
Copyright (c) 2014-2025 The GoCryptoTrader Developers.

GitHub: https://github.com/thrasher-corp/gocryptotrader
Kanban: https://github.com/orgs/thrasher-corp/projects/3
Slack:  https://join.slack.com/t/gocryptotrader/shared_invite/enQtNTQ5NDAxMjA2Mjc5LTc5ZDE1ZTNiOGM3ZGMyMmY1NTAxYWZhODE0MWM5N2JlZDk1NDU0YTViYzk4NTk3OTRiMDQzNGQ1YTc4YmRlMTk
Issues: https://github.com/thrasher-corp/gocryptotrader/issues

2025/01/28 10:40:52 Loading config file /Users/scottg/.gocryptotrader/config.json..
exchange CoinbasePro authenticated API support disabled due to default/empty APIKey/Secret/ClientID values
Logger initialised.
```

File: .github/workflows/tests.yml [hunk 10]
```diff
@@ -13,22 +13,32 @@ jobs:
...
             goarch: amd64
             psql: true
             skip_wrapper_tests: true
+            sonic: false
           - os: windows-latest
             goarch: amd64
             psql: true
             skip_wrapper_tests: true
+            sonic: false
+          - os: ubuntu-latest
```
Comment (gloriousCode) [inline .github/workflows/tests.yml:37]: I'm not seeing which one is sonic in anyway from the build list.
image
Or is there two that use sonic (the last two ubuntu builds)? Am I blind? 🙈

**Review by gloriousCode — CHANGES_REQUESTED** [review]
My comments feel very petty, but I'm happy overall with the work you've done

File: .github/workflows/tests.yml (hunk 10 above)
Comment (shazbert) [reply .github/workflows/tests.yml:37]: The last one is a docker build so it only comes up with false true, but the matrix build above, seems to be the only one with sonic set to true, but in the logs, it doesn't seem to pull the bytedance package down. I am trying to fix that now.

Comment (shazbert) [reply .github/workflows/tests.yml:37]: image
Unintended catch by you sir 🚀 but that should be good to go now.

Comment (gloriousCode) [reply .github/workflows/tests.yml:37]: Looks like the docker build is ded, but otherwise, looks good

Comment (shazbert) [reply .github/workflows/tests.yml:37]: Do you want me to fix the docker build in this PR? @gloriousCode

**Review by gloriousCode — APPROVED** [review]
tACK!

File: .github/workflows/tests.yml (hunk 10 above)
Comment (gloriousCode) [reply .github/workflows/tests.yml:37]: Sorry for blaming you for it.

**Review by thrasher- — CHANGES_REQUESTED** [review]
Makefile, GHA config and general functionality all looks good! Spotted this test error on it though:

https://github.com/thrasher-corp/gocryptotrader/actions/runs/13253592337/job/36996357960#step:11:4906

**Comment by shazbert** [comment]
> Makefile, GHA config and general functionality all looks good! Spotted this test error on it though:
> 
> https://github.com/thrasher-corp/gocryptotrader/actions/runs/13253592337/job/36996357960#step:11:4906

[fixed](https://github.com/thrasher-corp/gocryptotrader/pull/1623/commits/a843b585c9feedc21450ef3277cfe080b3b0dedf) good spotting!

File: .github/workflows/tests.yml [hunk 11]
```diff
@@ -84,13 +94,20 @@ jobs:
         echo "CGO_ENABLED=1" >> $GITHUB_ENV
       shell: bash
 
+    - name: Set test flags
+      run: |
+        if [ "${{ matrix.sonic }}" = "true" ]; then
+          echo "TEST_FLAGS=-tags sonic" >> $GITHUB_ENV
```
Comment (gbjk) [inline .github/workflows/tests.yml:100]: Shouldn't this just be [GOFLAGS](https://pkg.go.dev/cmd/go#hdr-Compile_packages_and_dependencies) env var that's supported out of the box ?
Note that we should *add* to it, not override it

File: .github/workflows/tests.yml [hunk 12]
```diff
@@ -84,13 +94,20 @@ jobs:
...
+        fi
+      shell: bash
+
     - name: Test
       run: | # PGSERVICEFILE isn't supported by lib/pq and will cause a panic if set
         unset PGSERVICEFILE
         if [ "${{ matrix.goarch }}" = "386" ]; then
-          go test -coverprofile coverage.txt -covermode atomic ./...
+          go test $TEST_FLAGS -coverprofile coverage.txt -covermode atomic ./...
```
Comment (gbjk) [inline .github/workflows/tests.yml:108]: This goes away with GOFLAGS env var.

**Comment by gbjk** [comment]
> Are you suggesting to default the main code to sonic out of the box not just the workflow tests?

Yes

File: Makefile [hunk 13]
```diff
@@ -78,3 +78,7 @@ endef
 check-jq:
 	@printf "Checking if jq is installed... "
 	@command -v jq >/dev/null 2>&1 && { printf "OK\n"; } || { printf "FAILED. Please install jq to proceed.\n"; exit 1; }
+
+.PHONY: sonic
+sonic:
+	go build -tags=sonic ./...
```
Comment (gbjk) [inline Makefile:84]: This is different from all previous examples.
You're not building sonic. You're building gct with sonic.
And it leaves a gap where install should be.

I suggest that users need to use `make build` and `install` with an argument:
```
make install JSON=sonic
```

Or, if we go the route I'm proposing:
```
make install JSON=golang
```

Also: Given the recent highlighting of `act`, I wonder if we remove Makefile and update docs to use `act`, and make that a (non-runtime) dependency for installing GCT from source.
Note: As we approach semver (right?), we should be issuing binaries anyway

File: LICENSE [hunk 14]
```diff
@@ -1,6 +1,6 @@
 The MIT License (MIT)
 
-Copyright (c) 2014-2024 The GoCryptoTrader Developers
+Copyright (c) 2014-2025 The GoCryptoTrader Developers
```
Comment (gbjk) [inline LICENSE:3]:
```suggestion
Copyright 2014 The GoCryptoTrader Developers
```
As per google, microsoft and linux. Lots of citations and links [here](https://softwareengineering.stackexchange.com/questions/210472/is-renewal-of-mit-license-needed-on-github-at-the-beginning-of-each-year)

Comment (shazbert) [reply LICENSE:3]: I don't mind having a single year. Will get more perspective on it. @thrasher- @gloriousCode

File: Makefile (hunk 13 above)
Comment (shazbert) [reply Makefile:84]: If we are going to allow customisation, I think a more general approach `make build FLAGS="-tags=sonic"` might be better. This gives users the flexibility to specify any build tags or flags they need without hardcoding specific options in the Makefile. If users have to manually specify flags every time, they might as well use go build or go install directly. The goal of make is to simplify workflows, not add unnecessary complexity.

We could:

Keep the Makefile but allow FLAGS to be overridden for advanced use cases.
Provide default targets like make build-sonic for common configurations.

As for act that is an option. But I have little to no experience in semver and issuing binaries so you can discuss with @thrasher- && @gloriousCode

**Comment by shazbert** [comment]
> > Are you suggesting to default the main code to sonic out of the box not just the workflow tests?
> 
> Yes

Ok cool,

> I propose that we're being too cautious, and creating unnecessary split code paths and feature support. I completely understand why it's reassuring to dip our toes in gently. But this either works right (and it probably does), or it doesn't.

I guess it comes from GCT library (as the more we update this library the further it drifts away from this and I could be completely wrong) wants to be as close to using internal dependencies as much as possible as default and all external dependencies are opt in. You're right in that this approach is being too cautious but I like that. 

Also it had to do with sonics different configurations Default, Fastest, Standard(aligns most closely to encoding/json). I was thinking we could have different build tags for those in future if I or somebody else ever gets around to it (other things are way more priority). Also we could hot swap in different JSON libraries if they were similar if they provide a certain use case.

Anyways this is my perspective and I am not against sonic being default either. @thrasher- @gloriousCode for your input

File: Makefile (hunk 13 above)
Comment (gbjk) [reply Makefile:84]: If you're going to use FLAGS, then:
```
GOFLAGS="-tags=hedgehogless" make build
```
makes more sense than adding it to the end, and means no extra support for FLAGS in Makefile.
Note: I'll accept nosonic or sonicless )

act: I've only just started using it

Semver is a separate topic, and issuing binaries doesn't need to be related to it, now that I think about it.

File: encoding/json/json.go [hunk 15]
```diff
@@ -0,0 +1,23 @@
+//go:build !sonic
+
+package json
+
+import "encoding/json" //nolint:depguard // This is a wrapper package for encoding/json. It serves as the default JSON package for GCT (GoCryptoTrader). All uses of JSON throughout the application should refer to this package with the default build.
+
+// Version is a constant string that represents the current JSON implementation package
+const Version = "encoding/json"
```
Comment (gbjk) [inline encoding/json/json.go:8]: I don't think Version is the right term for this.

Implementation maybe.

However I don't see any reason to need to know apart from logging it, which I'm about to recommend removing :)

File: main.go [hunk 16]
```diff
@@ -140,6 +141,8 @@ func main() {
 
 	gctscript.Setup()
 
+	gctlog.Infof(gctlog.Global, "JSON encoding is set to package `%s`", json.Version)
```
Comment (gbjk) [inline main.go:144]: I don't believe this is important at an Info level.
I actually don't think we should log it even at verbose + Debug.

**Review by gbjk — COMMENTED** [review]
Looks good 👍 Great work 🎉 

Only suggests again.

File: LICENSE (hunk 14 above)
Comment (thrasher-) [reply LICENSE:3]: You can leave this as is, additional context here https://opensource.stackexchange.com/questions/6389/how-do-the-years-specified-in-a-copyright-statement-work

BTC devs do the same thing https://github.com/bitcoin/bitcoin/blob/master/COPYING#L3 but isn't a big deal at the end of the day

Comment (gbjk) [reply LICENSE:3]: Sorry; My link was incorrect, in that it didn't jump to the desired answer, and actually it's the top link that I found most compelling.

Citations and links are [here](https://daniel.haxx.se/blog/2023/01/08/copyright-without-years/) and [here](https://hynek.me/til/copyright-years/).

Copyright isn't licensing, it's actually implicit anyway under the Berne convention.

[Golang](https://github.com/golang/go/blob/master/LICENSE) was one of the most compelling arguments for me.

Also as per your link, which I kinda feel makes my point too:
> Nowadays copyright notices and the year in the copyright notice are no longer very important, since copyright starts automatically when a creative work is created

Resolving conversation; Just wanted to articulate my point correctly.

Comment (thrasher-) [reply LICENSE:3]: It's fine, the link I provided shows all the necessary context anyway and thus my "Isn't a big deal" comment. So it can be left as is

File: encoding/json/json.go (hunk 15 above)
Comment (shazbert) [reply encoding/json/json.go:8]: I use this externally to confirm whether Sonic is in use, as I often mistype build tags. I’d prefer keeping it. I can change to implementation.

Comment (gloriousCode) [reply encoding/json/json.go:8]: I requested it to have an in-application acknowledgement of how it was built since that can happen as a seperate step. Kinda like how we log runtime args too. Can be helpful on log history.

File: main.go (hunk 16 above)
Comment (shazbert) [reply main.go:144]: The intent is just to show that the JSON implementation can be configured. Would it make more sense to log only when a non-default implementation is used and changed to debug?

Comment (gbjk) [reply main.go]: Up to you. Resolve if you're happy with it as is 🙂 
I always try to keep Info limited so that when I look at logs they don't overwhelm with information which isn't relevant.

**Review by gbjk — APPROVED** [review]
Looks good.

If you take the suggestion to make sonic the default, then I'll re-review that, obviously.

File: main.go (hunk 16 above)
Comment (shazbert) [reply main.go]: after consideration I am resolving, and add more pain to your log output!

File: Makefile (hunk 13 above)
Comment (thrasher-) [reply Makefile:84]: 1) Can keep the Makefile, as I find other recipes useful locally like pprof and only use act for simulating GHA workflows which I don't think it can do that also. We also want to keep the number of dependencies (runtime and non-runtime) down as much as possible for GCT ouf of the box. 
2) I'm happy with supporting both options who want a quick shorthand recipe and make install/make build param for power users (we do this with LTC where things are diisabled by NO_X).

File: .github/workflows/tests.yml [hunk 17]
```diff
@@ -13,22 +13,32 @@ jobs:
             goarch: amd64
```
Comment (gloriousCode) [inline .github/workflows/tests.yml:13]: I'm going to go ahead and leave a _code comment_ because then people can actually reply to eachother instead of big texts blocks that no one can reply to 🌈 

Before we learned that this json package was compatible with more targets than initially thought, the design and default options for this PR was well considered by shazbert. Now that it has changed, with arm64, linux & mac support, we can actually switch the default to using sonic. 
So it wasn't about dipping toes, shazbert had a windows/linux amd64 only pr and ensured compatibility for everyone else.

So I'd be happy for it to switch, so long as the builds default to sonic enabled with one for sonic disabled to ensure we catch anything improper and then decide from there whether we want to be a

**Comment by thrasher-** [comment]
> This is a comment only about the overall ideaology
> 
> I see from the tests that we are defaulting to sonic off, and we're testing amd64 with sonic both on and off.
> 
> I propose that we're being too cautious, and creating unnecessary split code paths and feature support. I completely understand why it's reassuring to dip our toes in gently. But this either works right (and it probably does), or it doesn't.
> 
> I'm _assuming_ our main production use cases are amd64/arm64 on linux, and we should optimise efficiency and ease of use for that main use-case. Guys, please correct that assumption and this proposal can die immediately. I'm assuming windows should be edge case for runtime/production.
> 
> Proposal / Endorsement:
> 
> * We should default to sonic on
> * We should refuse to run with sonic off if it _should_ be available according to the platform
> * We then have only have edge case tests where we turn sonic off

We can default sonic to on, since we're already using another json package jsonparser by default in many exchanges and other systems now.
We can have one job with it disabled to catch potential issues (linux amd64).
The makefile can therefore be adjusted to support a NO_SONIC build param if people want to build with native JSON support instead.

File: cmd/documentation/root_templates/root_readme.tmpl [hunk 18]
```diff
@@ -78,6 +78,7 @@ However, we welcome pull requests for any exchange which does not match this cri
 + WebGUI (discontinued).
 + Exchange HTTP mock testing. See [mock](/exchanges/mock/README.md).
 + Exchange multichain deposits and withdrawals for specific exchanges. See [multichain transfer support](/docs/MULTICHAIN_TRANSFER_SUPPORT.md).
++ Sonic JSON Integration. Revert back to the default Go 'encoding/json' package from the sonic library using go build tags `-tags=sonic_off`, by using the make file `make no_sonic`, or when compiling for a 32-bit architecture (GOARCH=386).
```
Comment (gbjk) [inline cmd/documentation/root_templates/root_readme.tmpl:81]: * Still don't think `Current Features` is the right place for this.. Sonic isn't a feature; it's an implementation choice.
* Even if this does stay here, "Revert back..." instructions is definitely not right here.

I suggest this gets moved to part of `Compiling instructions`

File: Makefile [hunk 19]
```diff
@@ -79,6 +79,6 @@ check-jq:
 	@printf "Checking if jq is installed... "
 	@command -v jq >/dev/null 2>&1 && { printf "OK\n"; } || { printf "FAILED. Please install jq to proceed.\n"; exit 1; }
 
-.PHONY: sonic
-sonic:
-	go build -tags=sonic ./...
\ No newline at end of file
+.PHONY: no_sonic
+no_sonic:
```
Comment (gloriousCode) [inline Makefile:83]:
```
make no_sonic 
go build -tags=sonic_off ./...
butts@butts-Laptop gocryptotrader % gocryptotrader 

   ______        ______                     __        ______                  __
  / ____/____   / ____/_____ __  __ ____   / /_ ____ /_  __/_____ ______ ____/ /___   _____
 / / __ / __ \ / /    / ___// / / // __ \ / __// __ \ / /  / ___// __  // __  // _ \ / ___/
/ /_/ // /_/ // /___ / /   / /_/ // /_/ // /_ / /_/ // /  / /   / /_/ // /_/ //  __// /
\____/ \____/ \____//_/    \__, // .___/ \__/ \____//_/  /_/    \__,_/ \__,_/ \___//_/
                          /____//_/
GoCryptoTrader v0.1 arm64 go1.23.5 pre-release.
This version is pre-release and is not intended to be used as a production ready trading framework or bot - use at your own risk.
Copyright (c) 2014-2025 The GoCryptoTrader Developers.

GitHub: https://github.com/thrasher-corp/gocryptotrader
Kanban: https://github.com/orgs/thrasher-corp/projects/3
Slack:  https://join.slack.com/t/gocryptotrader/shared_invite/enQtNTQ5NDAxMjA2Mjc5LTc5ZDE1ZTNiOGM3ZGMyMmY1NTAxYWZhODE0MWM5N2JlZDk1NDU0YTViYzk4NTk3OTRiMDQzNGQ1YTc4YmRlMTk
Issues: https://github.com/thrasher-corp/gocryptotrader/issues

2025/02/20 13:41:17 Loading config file /Users/butts/.gocryptotrader/config.json..
[INFO]  | LOG | 20/02/2025 13:41:17 | Logger initialised.
[INFO]  | LOG | 20/02/2025 13:41:17 | JSON encoding is set to package `bytedance/sonic`
```
This doesn't appear correct, but I'm no doctor

File: encoding/json/json.go [hunk 20]
```diff
@@ -1,8 +1,8 @@
-//go:build !sonic
+//go:build sonic_off || 386
 
 package json
 
-import "encoding/json" //nolint:depguard // This is a wrapper package for encoding/json. It serves as the default JSON package for GCT (GoCryptoTrader). All uses of JSON throughout the application should refer to this package with the default build.
+import "encoding/json" //nolint:depguard // This is a wrapper package for encoding/json. All uses of JSON throughout the application should refer to this when sonic_off build tag is used.
```
Comment (gbjk) [inline encoding/json/json.go:5]: * Seems like we're using the nolint as the package comment across the 3 json/*.go files.
* `encoding/json` is ambiguous in the comment, since it's referring to `golang.org/encoding/json` but inside `gct/encoding/json`
* I don't think we need to say that all uses of json throughout should refer to it

So I think we want a single package comment (common.go?) and then package comments in the other two relevant to them, and then the depguard nolint comment just explains it:
```
// json is an abstraction middleware package to allow switching between json encoder/decoder implementations
// The default implementation is sonic.
// Build with `sonic_off` or `386` tags to switch to golang.org/encoding/json.
```
and
```
import "encoding/json" //nolint:depguard // Acceptable use in gct json wrapper
 ```

**Review by gbjk — CHANGES_REQUESTED** [review]
Great work 🎉 

Only a couple of minor things.

File: Makefile (hunk 19 above)
Comment (shazbert) [reply Makefile:83]: [fixed hopefully](https://github.com/thrasher-corp/gocryptotrader/pull/1623/commits/ce013760b01dafa717e29345479bbca224225faa)

File: cmd/documentation/root_templates/root_readme.tmpl (hunk 18 above)
Comment (shazbert) [reply cmd/documentation/root_templates/root_readme.tmpl]: [cya mate](https://github.com/thrasher-corp/gocryptotrader/pull/1623/commits/d49cf8e43b9e167cce14b4b6cacb7dbf5637fdf5)

File: encoding/json/json.go (hunk 20 above)
Comment (shazbert) [reply encoding/json/json.go]: [done](https://github.com/thrasher-corp/gocryptotrader/pull/1623/commits/fcc0b27e2ccc9a6884f8fb29095e98fa743bd13c)

File: README.md [hunk 21]
```diff
@@ -101,15 +101,11 @@ When submitting a PR, please abide by our coding guidelines:
 + Code must adhere to our [coding style](https://github.com/thrasher-corp/gocryptotrader/blob/master/.github/CONTRIBUTING.md).
 + Pull requests need to be based on and opened against the `master` branch.
 
-## Compiling instructions
+## Compilation and run instructions
```
Comment (gbjk) [inline README.md:103]: Super-Duper Nit: "Compilation" tastes weird
``` Compiling and Running```

**Review by gbjk — APPROVED** [review]
Approved

File: README.md (hunk 21 above)
Comment (gbjk) [reply README.md:103]: Considered also splitting sections into `Compiling` and `Running`

File: Makefile (hunk 19 above)
Comment (gloriousCode) [reply Makefile:83]: Doesn't work for me. Is it working for you? Am I bad at `make`?

File: README.md (hunk 21 above)
Comment (shazbert) [reply README.md]: [done sir](https://github.com/thrasher-corp/gocryptotrader/pull/1623/commits/00ebc202a653e6722b7d9a6512a9c71bd70fd319)

File: Makefile (hunk 19 above)
Comment (shazbert) [reply Makefile:83]: maybe I am 😆 ?

**Review by gloriousCode — APPROVED** [review]
tACK

**Comment by thrasher-** [comment]
Just a note here: 
Defaulting to this on may slow down our migration to 1.24 and future Go versions. If they take a week or so to migrate to the latest go major version then that is fine. If however, they take months then we'll switch to defaulting off as our general turn around time for PR for the past few releases is days as we like to take advantage of all the new Go features as soon as they come available. I've asked for a ETA on when they're going to merge so this will be a good test to see: https://github.com/bytedance/sonic/pull/741
